server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java

```diff
@@ -698,13 +698,21 @@ public void run()
                                          .withEmitter(emitter)
                                          .withBalancerStrategyFactory(factory)
                                          .build();
-
-        for (DruidCoordinatorHelper helper : helpers) {
-          // Don't read state and run state in the same helper otherwise racy conditions may exist
-          if (leader && startingLeaderCounter == leaderCounter) {
-            params = helper.run(params);
+        try {
```

Comment:
you might want to use try with resources with the factory, just in case an exception is thrown earlier. This will also close the factory automatically.